src/roadveh_cmd.cpp

```diff
@@ -768,8 +768,16 @@ struct OvertakeData {
 static Vehicle *EnumFindVehBlockingOvertake(Vehicle *v, void *data)
 {
 	const OvertakeData *od = (OvertakeData*)data;
-
-	return (v->type == VEH_ROAD && v->First() == v && v != od->u && v != od->v) ? v : nullptr;
+	TileIndexDiff    front = TileOffsByDiagDir(DirToDiagDir(od->v->direction));
+
+	/* Distinct road engine is blocking if it's within the tile or the next few of the passer as well as either it being
+	 * directionally opposed to passer, the passed pending movement, or the blocker being stopped while overtaking. */
+	return (v->type == VEH_ROAD && v->First() == v && v != od->u && v != od->v
```

Comment:
This should be broken up into multiple returns for readability, preferably with comments explaining each case. The ternary at the end is particularly confusing.